jetty-core/jetty-ee/src/main/java/org/eclipse/jetty/ee/WebAppClassLoader.java

```diff
@@ -0,0 +1,657 @@
+import org.eclipse.jetty.util.resource.ResourceCollators;
+import org.eclipse.jetty.util.resource.ResourceFactory;
+import org.eclipse.jetty.util.resource.Resources;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * ClassLoader for Http Context.
+ * <p>
+ * Specializes URLClassLoader with some utility and file mapping
+ * methods.
+ * <p>
+ * This loader implements the Servlet specification behavior that may invert the normal Java classloader parent priority
+ * behaviour. The {@link ClassVisibilityChecker} API of the {@link WebAppClassLoader.Context} implementation is
+ * used to determine which classes from the parent classloader are hidden from the context, and which are protected
+ * from being overridden by the context.
+ * <p>
+ * Java compliant loading, where the parent loader always has priority, can be selected with the
+ * {@link Context#isParentLoaderPriority()} method.
+ */
+@ManagedObject
+public class WebAppClassLoader extends URLClassLoader implements ClassVisibilityChecker
+{
+    static
+    {
+        registerAsParallelCapable();
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(WebAppClassLoader.class);
+    private static final ThreadLocal<Boolean> __loadHiddenClasses = new ThreadLocal<>();
+
+    private final Context _context;
+    private final ClassLoader _parent;
+    private final Set<String> _extensions = new HashSet<>();
+    private String _name = String.valueOf(hashCode());
+    private final List<ClassFileTransformer> _transformers = new CopyOnWriteArrayList<>();
+    private final ResourceFactory.Closeable _resourceFactory = ResourceFactory.closeable();
+
+    /**
+     * The Context in which the classloader operates.
+     */
+    public interface Context extends ClassVisibilityChecker
+    {
+        /**
+         * Convert a URL or path to a Resource.
+         * The default implementation
+         * is a wrapper for {@link ResourceFactory#newResource(String)}.
+         *
+         * @param urlOrPath The URL or path to convert
+         * @return The Resource for the URL/path
+         * @throws IOException The Resource could not be created.
+         */
+        Resource newResource(String urlOrPath) throws IOException;
+
+        /**
+         * @return Returns the permissions.
+         */
+        PermissionCollection getPermissions();
+
+        /**
+         * @return True if the classloader should delegate first to the parent
+         * classloader (standard java behaviour) or false if the classloader
+         * should first try to load from WEB-INF/lib or WEB-INF/classes (servlet
+         * spec recommendation).
+         */
+        boolean isParentLoaderPriority();
+
+        List<Resource> getExtraClasspath();
+
+        boolean isHiddenResource(String name, URL parentUrl);
+
+        String[] getHiddenClasses();
+
+        ClassMatcher getHiddenClassMatcher();
+
+        boolean isProtectedResource(String name, URL webappUrl);
```

Comment:
This previously had a default implementation of `return false;`

Reply:
looking at this, a lot of these can be defaulted, not just that one method.
let me give it a go.